flow/model/model.go

```diff
@@ -339,39 +337,31 @@ func NewCDCRecordStream() *CDCRecordStream {
 		emptySignal:            make(chan bool, 1),
 		RelationMessageMapping: make(chan *RelationMessageMapping, 1),
 		lastCheckpointSet:      false,
-		lastCheckPointID:       0,
-		firstCheckPointID:      0,
+		lastCheckPointID:       atomic.Int64{},
+		firstCheckPointID:      atomic.Int64{},
 	}
 }
 
 func (r *CDCRecordStream) UpdateLatestCheckpoint(val int64) {
-	r.checkpointMutex.Lock()
-	defer r.checkpointMutex.Unlock()
+	r.firstCheckPointID.CompareAndSwap(0, val)
 
-	if r.firstCheckPointID == 0 {
-		r.firstCheckPointID = val
-	}
-
-	if val > r.lastCheckPointID {
-		r.lastCheckPointID = val
+	// TODO update with https://github.com/golang/go/issues/63999 once implemented
+	// r.lastCheckPointID.Max(val)
+	oldLast := r.lastCheckPointID.Load()
+	for oldLast < val && !r.lastCheckPointID.CompareAndSwap(oldLast, val) {
+		oldLast = r.lastCheckPointID.Load()
 	}
 }
 
 func (r *CDCRecordStream) GetFirstCheckpoint() int64 {
-	r.checkpointMutex.Lock()
-	defer r.checkpointMutex.Unlock()
-
-	return r.firstCheckPointID
+	return r.firstCheckPointID.Load()
 }
 
 func (r *CDCRecordStream) GetLastCheckpoint() (int64, error) {
-	r.checkpointMutex.Lock()
-	defer r.checkpointMutex.Unlock()
-
 	if !r.lastCheckpointSet {
```

Comment:
`lastCheckpointSet` seems kind of funny since it gets written to by `Close` without synchronization. Seems like this struct has a mode between collection/read which would be cleaner as a channel which gets sent first/last checkpoint id at end of pull or something

Reply:
I agree that channels to communicate most fields in this struct would be a good idea. I only wish channels had a GetOrWait primitive :-)